src/core/qgsmaplayerlegend.cpp

```diff
@@ -206,6 +207,16 @@ QList<QgsLayerTreeModelLegendNode*> QgsDefaultVectorLayerLegend::createLayerTree
   if ( nodes.count() == 1 && nodes[0]->data( Qt::EditRole ).toString().isEmpty() )
     nodes[0]->setEmbeddedInParent( true );
 
+
+  if ( mLayer->diagramRenderer() )
```

Comment:
done. You're right, I avoid having the lengend appear even if the diagram is disabled. Thanks @nyalldawson